Detectors/ITSMFT/ITS/tracking/cuda/src/ClusterLinesGPU.cu

```diff
@@ -0,0 +1,287 @@
+                  determinantSecond;
+
+  mBMatrix[0] =
+    (firstLine.cosinesDirector[1] * covarianceFirst[2] * (-firstLine.cosinesDirector[1] * firstLine.originPoint[0] + firstLine.cosinesDirector[0] * firstLine.originPoint[1]) +
+     firstLine.cosinesDirector[2] * covarianceFirst[1] * (-firstLine.cosinesDirector[2] * firstLine.originPoint[0] + firstLine.cosinesDirector[0] * firstLine.originPoint[2])) /
+    determinantFirst;
+
+  mBMatrix[0] +=
+    (secondLine.cosinesDirector[1] * covarianceSecond[2] * (-secondLine.cosinesDirector[1] * secondLine.originPoint[0] + secondLine.cosinesDirector[0] * secondLine.originPoint[1]) +
+     secondLine.cosinesDirector[2] * covarianceSecond[1] *
+       (-secondLine.cosinesDirector[2] * secondLine.originPoint[0] +
+        secondLine.cosinesDirector[0] * secondLine.originPoint[2])) /
+    determinantSecond;
+
+  mBMatrix[1] =
+    (firstLine.cosinesDirector[0] * covarianceFirst[2] * (-firstLine.cosinesDirector[0] * firstLine.originPoint[1] + firstLine.cosinesDirector[1] * firstLine.originPoint[0]) +
+     firstLine.cosinesDirector[2] * covarianceFirst[0] * (-firstLine.cosinesDirector[2] * firstLine.originPoint[1] + firstLine.cosinesDirector[1] * firstLine.originPoint[2])) /
+    determinantFirst;
+
+  mBMatrix[1] +=
+    (secondLine.cosinesDirector[0] * covarianceSecond[2] * (-secondLine.cosinesDirector[0] * secondLine.originPoint[1] + secondLine.cosinesDirector[1] * secondLine.originPoint[0]) +
+     secondLine.cosinesDirector[2] * covarianceSecond[0] *
+       (-secondLine.cosinesDirector[2] * secondLine.originPoint[1] +
+        secondLine.cosinesDirector[1] * secondLine.originPoint[2])) /
+    determinantSecond;
+
+  mBMatrix[2] =
+    (firstLine.cosinesDirector[0] * covarianceFirst[1] * (-firstLine.cosinesDirector[0] * firstLine.originPoint[2] + firstLine.cosinesDirector[2] * firstLine.originPoint[0]) +
+     firstLine.cosinesDirector[1] * covarianceFirst[0] * (-firstLine.cosinesDirector[1] * firstLine.originPoint[2] + firstLine.cosinesDirector[2] * firstLine.originPoint[1])) /
+    determinantFirst;
+
+  mBMatrix[2] +=
+    (secondLine.cosinesDirector[0] * covarianceSecond[1] * (-secondLine.cosinesDirector[0] * secondLine.originPoint[2] + secondLine.cosinesDirector[2] * secondLine.originPoint[0]) +
+     secondLine.cosinesDirector[1] * covarianceSecond[0] *
+       (-secondLine.cosinesDirector[1] * secondLine.originPoint[2] +
+        secondLine.cosinesDirector[2] * secondLine.originPoint[1])) /
+    determinantSecond;
+
+  computeClusterCentroid();
+}
+
+GPUd() ClusterLinesGPU::ClusterLinesGPU(const int& firstLabel, const Line& firstLine, const int& secondLabel, const Line& secondLine,
+                                        const unsigned char weight) : mNContributors{2}
+{
+  mLabels[0] = firstLabel;
+  mLabels[1] = secondLabel;
+
+  float covarianceFirst[3];
+  float covarianceSecond[3];
+
+  for (int i{0}; i < 3; ++i) {
+    covarianceFirst[i] = 1.f;
```

Comment:
as far as I understand, the covariances are placeholders at the moment, in case you decide in the future to use real errors derived from cluster errors. But I am wondering if it makes sense at all to use errors (they will be in 5-10 microns range at the vertex) given that the multiple scattering and neglected bending (for xy) is anyway larger?

Reply:
This was indeed done to include in the computation also those calculations, to be fair with benchmarks. Indeed what you say is also interesting. Will be useful to be taken into account in next optimizations.